### PR TITLE
[api] Separate loglistener from logfilter

### PR DESCRIPTION
**Pull request by Liuhaai**
# Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:

- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Liuhaai** [comment]
b/c other pr depends on this one. Ranaming job is scheduled in the future https://github.com/iotexproject/iotex-core/issues/3295

File: api/coreservice.go [hunk 1]
```diff
@@ -127,12 +127,10 @@ type (
 		BlockMetas(start uint64, count uint64) ([]*iotextypes.BlockMeta, error)
 		// BlockMetaByHash returns blockmeta response by block hash
 		BlockMetaByHash(blkHash string) (*iotextypes.BlockMeta, error)
-		// LogsInBlock filter logs in the block x
-		LogsInBlock(filter *logfilter.LogFilter, blockNumber uint64) ([]*iotextypes.Log, error)
```
Comment (Liuhaai) [inline api/coreservice.go:131]: this func in the interface is useless

File: api/grpcserver.go [hunk 2]
```diff
@@ -394,21 +394,36 @@ func (svr *GRPCServer) GetLogs(ctx context.Context, in *iotexapi.GetLogsRequest)
...
-		ret, err = svr.coreService.LogsInBlockByHash(logfilter.NewLogFilter(in.GetFilter(), nil, nil), hash.BytesToHash256(in.GetByBlock().BlockHash))
+		blkHash := hash.BytesToHash256(in.GetByBlock().BlockHash)
+		logs, err = svr.coreService.LogsInBlockByHash(logfilter.NewLogFilter(in.GetFilter()), blkHash)
+		if err != nil {
+			break
+		}
+		hashes = make([]hash.Hash256, 0, len(logs))
+		for range logs {
+			hashes = append(hashes, blkHash)
+		}
```
Comment (CoderZhi) [inline api/grpcserver.go:411]: why not compose `retval` in this for loop directly?

**Review by CoderZhi — APPROVED** [review]

File: api/coreservice_test.go [hunk 3]
```diff
@@ -73,12 +75,10 @@ func TestLogsInRange(t *testing.T) {
 		to, err := strconv.ParseUint(testData.ToBlock, 10, 64)
 		require.NoError(err)
 
-		logs, err := svr.web3Server.coreService.LogsInRange(logfilter.NewLogFilter(&filter, nil, nil), from, to, uint64(0))
+		_, _, err = svr.web3Server.coreService.LogsInRange(logfilter.NewLogFilter(&filter), from, to, uint64(0))
 		expectedErr := errors.New("invalid start and end height")
-		expectedValue := []*iotextypes.Log(nil)
 		require.Error(err)
 		require.Equal(expectedErr.Error(), err.Error())
-		require.Equal(expectedValue, logs)
```
Comment (dustinxie) [inline api/coreservice_test.go:81]: change to `require.Zero(len(logs))`

File: api/coreservice_test.go [hunk 4]
```diff
@@ -89,12 +89,10 @@ func TestLogsInRange(t *testing.T) {
 		to, err := strconv.ParseUint(testData.ToBlock, 10, 64)
 		require.NoError(err)
 
-		logs, err := svr.web3Server.coreService.LogsInRange(logfilter.NewLogFilter(&filter, nil, nil), from, to, uint64(0))
+		_, _, err = svr.web3Server.coreService.LogsInRange(logfilter.NewLogFilter(&filter), from, to, uint64(0))
 		expectedErr := errors.New("start block > tip height")
-		expectedValue := []*iotextypes.Log(nil)
 		require.Error(err)
 		require.Equal(expectedErr.Error(), err.Error())
-		require.Equal(expectedValue, logs)
```
Comment (dustinxie) [inline api/coreservice_test.go:97]: same here

File: api/logfilter/logfilter.go [hunk 5]
```diff
@@ -39,55 +33,23 @@ type LogFilter struct {
...
+	res := []*action.Log{}
 	for _, r := range receipts {
-		for _, v := range r.Logs() {
-			log := v.ConvertToLogPb()
-			if l.match(log) {
-				log.BlkHash = blkHash[:]
-				logs = append(logs, log)
+		for j, v := range r.Logs() {
+			if l.match(v.ConvertToLogPb()) {
+				res = append(res, r.Logs()[j])
```
Comment (dustinxie) [inline api/logfilter/logfilter.go:48]: can get rid of `v`
```
logs := r.Logs()
for i := range logs {
    \\ use logs[i] here
}
```

File: api/loglistener.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
+package api
+
+import (
+	"github.com/iotexproject/iotex-proto/golang/iotexapi"
+	"go.uber.org/zap"
+
+	logfilter "github.com/iotexproject/iotex-core/api/logfilter"
```
Comment (dustinxie) [inline api/loglistener.go:7]: `logfilter` at beginning is not needed?

File: api/coreservice.go [hunk 7]
```diff
@@ -1372,62 +1370,53 @@ func (core *coreService) reverseActionsInBlock(blk *block.Block, reverseStart, c
...
 		return nil, err
 	}
 
-	h, err := core.dao.GetBlockHash(blockNumber)
-	if err != nil {
-		return nil, err
-	}
-
-	return filter.MatchLogs(receipts, h), nil
+	return filter.MatchLogs(receipts), nil
```
Comment (dustinxie) [inline api/coreservice.go:1396]: better fill in `log.BlkHash` here, b/c `logsInBlock()` is used by couple of other places

File: api/coreservice.go [hunk 8]
```diff
@@ -1372,62 +1370,53 @@ func (core *coreService) reverseActionsInBlock(blk *block.Block, reverseStart, c
...
 }
 
 // LogsInRange filter logs among [start, end] blocks
-func (core *coreService) LogsInRange(filter *logfilter.LogFilter, start, end, paginationSize uint64) ([]*iotextypes.Log, error) {
+func (core *coreService) LogsInRange(filter *logfilter.LogFilter, start, end, paginationSize uint64) ([]*action.Log, []hash.Hash256, error) {
```
Comment (dustinxie) [inline api/coreservice.go:1400]: don't need to change here at all, see above comment

**Review by dustinxie — APPROVED** [review]

File: api/grpcserver.go (hunk 2 above)
Comment (Liuhaai) [reply api/grpcserver.go:411]: To be solved in another pr

File: api/logfilter/logfilter.go (hunk 5 above)
Comment (Liuhaai) [reply api/logfilter/logfilter.go:48]: To be solved in another pr

File: api/loglistener.go (hunk 6 above)
Comment (Liuhaai) [reply api/loglistener.go:7]: To be solved in another pr

**Pull request merged**